### DISCUSSION TITLE
The sound of some games is having problems.

### DISCUSSION THREAD

**Discussion by sonht-0555** [thenick775/gbajs3#136] [General]
@thenick775 I'm trying to play the game Earthbound [1-2].gba.  The audio experience is having problems. Every time I touch the screen, there is a "popping" noise like a slight audio noise.  You can try wearing headphones and playing a few games, sometimes the sound will be interrupted, before using Module.setFastForwardMultiplier() this error did not appear.

**Comment by thenick775** [comment]
Could you attach a video with sound showcasing the problem?

**Reply by sonht-0555** [reply]
https://github.com/thenick775/gbajs3/assets/26568422/ac833ccd-2245-4871-90db-5de2e5e62e4c

**Reply by sonht-0555** [reply]
@thenick775 Please put on headphones, you will clearly feel the sound of the background music

**Reply by thenick775** [reply]
Do you only see this in power saver mode?

**Reply by thenick775** [reply]
This may make sense as the audio is being handled on the main thread and not in an audio worklet

**Reply by sonht-0555** [reply]
https://github.com/thenick775/gbajs3/assets/26568422/9a47bada-e5d3-42a9-b0df-819810f29dc0

@thenick775 Yup, even on normal battery, we still encounter that phenomenon

**Reply by thenick775** [reply]
Perfect, this is probably what my gut first told me.

Our audio is experiencing jitter because of the main thread actions taken when using controls or interacting with the UI.

I can try to muck with the buffers to see if that helps at all, but I think the main thread audio is the root of the problem.

Pursuing audioworklet support, or a fully threaded implementation with proxying to a pthread in emacripten may help.

**Reply by sonht-0555** [reply]
@thenick775  I think part of the reason is also due to the way the mgba core processes audio output, causing the audioworklet to return audio data is not good, even when doing nothing, the popping sounds will not be clear, instead  Is it because the emulator doesn't perform it well (I tested it with the [NBA](https://nealmm.github.io/web-gba-emu/) core the output sound is really clear when using audioworklet)

**Reply by sonht-0555** [reply]
`Module.setFastForwardMultiplier()` since using this module to overcome the iOS Savemode barrier, the problem is really obvious with the sound when we perform touch actions on the screen

**Reply by thenick775** [reply]
Can you link to the NBA code that is processing the audio buffers? If they are using audioworklets id be very curious to see how they are handling async buffer fills from a web worker.

My thoughts are it's not that mGBA is doing a bad job, it's that we're using SDL2 to process the sound.

SDL/emscripten in this case is using script processor node, which ends up running on the main thread since everything here is running in the main thread right now.

The problem with everything being in the main thread event loop is that's where user actions like moving controls also happens, pushing the audio buffer fills back a bit, causing the jitter we're hearing. Extending the audio buffers may be an option, but I feel this is a larger issue.

Without power saver mode on, the behavior should be identical to what we had before at regular speed in the emulator, so I bet if I load an older version of gbajs3 it'll also be apparent there.

My best thoughts are that it's not exactly related to the way we're skipping frames, but definitely becomes more apparent when doing so!

**Reply by thenick775** [reply]
Seems like they have threading and the memory issues sorted out, I think that'll be the way to go if I can replicate the same setup using threads in mGBA.

Of note, this means that you will have to have secure context , and corp/coep headers to be able to use the emulator core at all. This would potentially break static hosting as it exists today.

**Reply by sonht-0555** [reply]
@thenick775 wow, so much knowledge. I don't know about SDL2. Basically I just understand it handles the audio output for the emulator. Have you tried looking at the NBA code in the link above? I don't know if you have any solution

**Reply by sonht-0555** [reply]
Do you intend to return to the old version?

**Reply by thenick775** [reply]
I don't intent to, unless we can confirm that the problem was not present before we implemented the performance enhancements.

I do definitely plan to work on rectifying this issue though! 😎 

Are you able to confirm that the problem does not exist in the latest [staging](https://thenick775.github.io/gbajs3/) deploy?

I've reset it back to before the fast forward and frame skipping performance enhancements.

Mathematically, the problem should manifest in the same way in this deploy.

**Reply by sonht-0555** [reply]
@thenick775 https://github.com/thenick775/gbajs3/assets/26568422/d6c378a2-07f3-4d38-a097-c9f64c7dcd47

There is no audio interruption in this version. problem has been identified

**Reply by thenick775** [reply]
Awesome, I'll try to work this regression here soon!

**Reply by thenick775** [reply]
Additionally, as I'm getting more familiar with the NBA core, looks like it should be fairly easy to integrate it into this UI.

From its internals, it looks like it supports everything but screenshots, the glue code just has to be written to interact with it from js.

Since this is a one man project so far, I can't give a timeline, but I plan to add NBA to the core list while bringing back switchable cores!

**Reply by sonht-0555** [reply]
@thenick775 
Ya NBA, I'm really looking forward to it. Hope you will do this soon. 

If possible, can you create a Repo about NBA Wasm like [mGBA Wasm](https://github.com/thenick775/mgba/tree/feature/wasm)?   I'll try integrating it into the UI and see how it goes.

It's embarrassing because I don't know much about Wasm :( so I can't help you

**Reply by thenick775** [reply]
That is the plan as I free up and am available!

I'll let you know once my fork is in place 😎

**Reply by sonht-0555** [reply]
@thenick775 Very willing, if there's anything new you need me to test, please message me here.

**Reply by thenick775** [reply]
@sonht0555 are you able to check how the sound is doing in the latest [staging](https://thenick775.github.io/gbajs3/) deploy?

I think I may have found the issue with some floating point rounding in the cores main loop.

This should fix the problems when not in low power mode, at regular speed.

If this helps I can potentially apply a similar styled fix when the user is in low power mode (the problem was working to our advantage in that case).

Thanks for testing this by the way, I've had some trouble hearing since a recent car accident, so your help in testing this is definitely appreciated 😎

**Reply by YiLee01** [reply]
Hi, I want to ask, have you fixed the floating point rounding in the cores main loop issue? I noticed that sometimes when the action button is touched quickly, the game (I'm playing Pokemon Emerald) will have intermittent tearing sounds, I guess it may be the same problem that @sonht0555 encountered.

**Reply by thenick775** [reply]
Yep that's the same problem!

I have not pushed a full fix yet as I'm currently experimenting with different approaches, but feel free to test the staging link above without low power mode, and at regular speed and see if that helps as a test!

I will close #138 once completed.

I'm planning on getting a more holistic solution built out and sent to production soon!

**Reply by thenick775** [reply]
If your interested in the problem as a whole, this article sums it up very well:

https://gafferongames.com/post/fix_your_timestep/

**Reply by thenick775** [reply]
@YiLee01 if you could attach a video with sound showcasing what you're seeing/hearing, that would also be appreciated!

**Reply by YiLee01** [reply]
[staging](https://thenick775.github.io/gbajs3/) Where is the code branch where the link is located?

**Reply by sonht-0555** [reply]
https://github.com/thenick775/gbajs3/assets/26568422/487019b8-fe56-4033-bc8b-61953258389d

Sorry, I was a bit busy yesterday. The problem is still not fixed :((

**Reply by sonht-0555** [reply]
@thenick775

**Reply by YiLee01** [reply]
Hello, I sent you a video of the game at https://gba.nicholas-vancise.dev. I found that in low power mode, the frame rate of the game is significantly reduced, and when switching to low power mode, there will be obvious tearing sound. In addition, the tearing sound is particularly obvious when the game is fast forwarded.

https://github.com/thenick775/gbajs3/assets/50646519/a1ea7502-e6ee-4607-ba8a-347eb72c552e

**Reply by thenick775** [reply]
Thanks @sonht0555 !

**Reply by thenick775** [reply]
@YiLee01 the branch deployed to staging is arbitrary, but currently its this [branch](https://github.com/thenick775/gbajs3/compare/master...thenick775:gbajs3:async-controls-button-test) so I can test on a real device.

The code for the mGBA wasm core is located [here](https://github.com/thenick775/mgba/tree/feature/wasm).

The changes in the branch deployed to staging is dirty code from my local machine, but I'm messing with the [main loop](https://github.com/thenick775/mgba/blob/feature/wasm/src/platform/wasm/main.c#L70)

**Reply by thenick775** [reply]
The tearing when transitioning is hard to prevent, so I'm focusing on the frame skipping when the game is running, since everything is on the main loop running browser actions take time, and the main loop and control actions are "skipping an extra frame" in short because of the virtual control interactions.

The tearing when switching to low power mode is because the audio buffer gets filled twice, and we can't detect when a device goes into low power mode in a more ergonomic way yet!

**Reply by YiLee01** [reply]
In addition to manually entering low power mode from the control settings, sometimes some phones with slightly poorer configurations may also automatically enter a similar "low power mode" after running for a period of time due to the heat of the body. I found that in this case, the game does have visible frame skipping. Of course, it will be more obvious when manually controlling the settings to enter low power mode, which is also reflected in my video.
The above are some of the situations I observed and found, I hope it will be helpful to you.

**Reply by thenick775** [reply]
If you all want to try again, I may have a slightly better solution up in the current staging deploy.

There will still be some popping, but it _should_ be reduced, and it _should_ work slightly better with the current fast forward implementation.

I may push this one as a stop gap fix in the meantime if this helps in regular speed+no low power mode. but it might have to wait until I work out some kinks there.

I'd open this in an incognito tab to ensure that you have the most up to date staging version!

**Reply by thenick775** [reply]
If this solution isn't preferable, the only other immediate stop gap I see is to turn off any automatic frame skipping.

Meaning you'd have to manually turn fast forward to 2 to maintain  60/fps when low power mode is active, since we can't detect it through other means.

**Reply by sonht-0555** [reply]
@thenick775 looking forward. I will wait to test it. I also hope you don't remove the 60fps feature in battery save mode

**Reply by thenick775** [reply]
@sonht0555 check out this [staging deploy](https://thenick775.github.io/gbajs3/) when you have a chance, this may be about as good as I can get it in a general sense. I recommend opening this in a private tab, or close and refresh 5 times, to ensure that you pick up the latest changes.

I may implement a settings view to mess with these values before a game is loaded (like the original mGBA) as this may not work for all devices in all situations. I've avoided that so far as I want to try and keep the UI simple, and one-size-fits all, but that may not work in this situation across all devices and all games.

I think I've reduced the jitter from what I can currently hear.

Current items that have been changed:
- [x] audio is scaled when fast forwarding
- [x] frame rate rendering calculations have been improved
- [x] audio buffers have been resized and sample rate has been changed

Things that should still work:
- [x] fps detection for low power mode (will experience some audio jitter in low power mode)

Future items:
- [ ] settings menu where audio settings can be modified before starting a rom

I'll be merging this into main shortly as I think this is a general improvement, even if there are still some problems 😎

**Reply by sonht-0555** [reply]
@thenick775 I tested this version and noticed many changes. Major vibration errors have been fixed. Instead, I tried wearing headphones and tested harder, but the sound was still slightly jerky. If you play the game in headphone mode, you will still find that the sound is not clean and balanced, sometimes there will still be small noises when moving the button. In both normal mode and outside sound mode, we almost don't notice or hear these small popping sounds.

**Reply by thenick775** [reply]
Sweet, I'm going to close this for now then as that sounds successful to me!

More work can be done to hopefully minimize the small popping sounds, but that will most likely require threading, and that's not guaranteed to fix the issue 100%.

Hopefully the addition of a settings menu allows for better configuration per game in the future!

Glad to see that this was a decent improvement! 😎

Thanks for all your collaboration!

**Reply by YiLee01** [reply]
Thank you very much for your hard work in fixing the audio issues! ❤️ 
I tested it and noticed significant improvements; the fast-forward audio is much cleaner now. However, I still encountered some additional issues. The smoothness of the game visuals seems to have decreased, and I can perceive a drop in frame rate.

**Reply by thenick775** [reply]
@YiLee01 are you playing with low power mode on or with fast forward on when you noticed the frame rate drops?

Are you able to submit a video of what you're seeing? Happy to take a look 😎 

Some of the enhancements above surrounded avoiding running extra frame skips, in essence preferring to round down by 30 percent, this has a larger effect if you're playing in the modes specified above.

I've also added additional frame skipping when in fast forward modes, equal to the fast forward speed (1-5) in the last pr. This increased fast forward performance by a significant amount allowing for cleaner audio.

If in regular power mode and without fast forward that may be something I can fix if there's a reproducible example 😎

**Reply by thenick775** [reply]
Well I think that's my problem right there, it's most likely the extra frame skipping I added being off by 1.

~~@YiLee01 do you see this after you've adjusted fast forward but not before if you close and reopen the emulator?~~

**Reply by thenick775** [reply]
@YiLee01 let me know if this staging [deploy](https://thenick775.github.io/gbajs3/) fixes what your seeing without impacting the audio too much!

I recommend opening in an incognito tab or refreshing ~5 times to ensure you have the latest files through the service worker.

This fix may hit main pretty fast, depending on how testing goes here shortly.

I thought I saw something a bit strange when playing inclement emerald, but wouldn't have noticed without you pointing this out 🚀

**Reply by YiLee01** [reply]
> @YiLee01 let me know if this staging [deploy](https://thenick775.github.io/gbajs3/) fixes what your seeing without impacting the audio too much!
> 
> I recommend opening in an incognito tab or refreshing ~5 times to ensure you have the latest files through the service worker.
> 
> This fix may hit main pretty fast, depending on how testing goes here shortly.
> 
> I thought I saw something a bit strange when playing inclement emerald, but wouldn't have noticed without you pointing this out 🚀

I apologize for the delay in responding due to my busy schedule. I’m delighted to see how quickly you addressed the issue. I spent some time playing the game (not in low power mode) and noticed significant improvements in smoothness. The audio is excellent, and the frame rate is perfect, making for a very enjoyable experience. Thank you for your hard work! ❤️

However, I’ve encountered a few instances of the game freezing, particularly when saving in Pokémon games. The game freezes with half the dialogue box visible and becomes unresponsive, even when loading saved states. I’m still investigating the cause and will continue testing.

I hope this helps convey your feedback effectively!

**Reply by thenick775** [reply]
Thanks for the confirmations @YiLee01 on this issue!

I'll keep an eye out to see if I can reproduce the same thing in reference to freezing, does this happen after you background the web app perhaps? Additionally, which ROM(s) have you experienced this problem in?

If you can grab a video of this happening and/or console log output of any type indicating a problem, please open another issue so we can investigate and address! I'll do the same if I experience this problem as well 😎

**Reply by YiLee01** [reply]
Yes, I am planning to open a new issue. Currently, I am continuing to play and test the game to reproduce the problem. I have encountered the issue in Pokémon Gold, Silver, and Ruby. It might be a ROM issue, but if I determine it is a problem with the emulator, I will open a new issue and provide detailed information.

**Comment by thenick775** [comment]
Note: most problems fixed in #138 🚀

**Reply by thenick775** [reply]
Note: frame skip regression fixed in #148 🚀
